internal/cmd/gocdk/add.go

```diff
@@ -16,10 +16,100 @@ package main
 
 import (
 	"context"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
 
 	"golang.org/x/xerrors"
 )
 
+type portableTypeInfo struct {
+	name       string // the portable type name
+	goDemoPath string // the path to the .go file to add to the project for static.Open
+	demoURL    string // the URL for the demo
+}
+
+var portableTypes = []*portableTypeInfo{
+	{
+		name:       "blob.Bucket",
+		goDemoPath: "/demo/blob.bucket/demo_blob_bucket.go",
+		demoURL:    "/demo/blob.bucket/",
+	},
+}
+
+// TODO(rvangent): Add tests for add(), including for each supported portableType.
+
 func add(ctx context.Context, pctx *processContext, args []string) error {
-	return xerrors.New("not implemented")
+	// Compute a sorted slice of available portable types for usage.
+	var avail []string
+	for _, pt := range portableTypes {
+		avail = append(avail, pt.name)
+	}
+	sort.Strings(avail)
+	usageMsg := fmt.Sprintf("gocdk add <%s>", strings.Join(avail, "|"))
```

Comment:
We already know this list is going to be 5 or more, which seems like it would be unwieldy for the places this message would be used.

Reply:
What do you suggest for help then? The list needs to viewable somehow. On separate lines? Or a separate command?

Reply:
I mean, I think the final print you have is a fine way of showing it. The information seems appropriate if the type is misspelled or missing.

The problem with this usage message is that the list will be spilled out when passing a misspelled flag or passing too many arguments, which seems unfortunate to me.

Reply:
IMHO we need to take an overall look at how flag parsing and help works; we can fix/improve this as part of that effort.

For example, `--help` kind of sucks now; it shows the available flags but not the actual usage message (including non-flag arguments):

```
$ gocdk add-ptype --help
Usage of add:
  -force
    	re-add even the portable type even if it has already been added, overwriting previous files
  -help
    	displays this help message
```

while using a wrong flag shows the full usage, but not the available flags:


```
$ gocdk add-ptype --unknownflag
usage: gocdk add <blob.Bucket>: flag provided but not defined: -unknownflag
`

Reply:
Filed https://github.com/google/go-cloud/issues/1997 to follow u.

Reply:
I agree. I've opened #1991 to look at the flag parsing one and #1810 tracks having a more comprehensive help command. My tendency is to try to address these sort of things in a cross-cutting way once we know what we need. I've added details to both issues with links to code on how I addressed these for gg, my Git wrapper that bumped into most of these same problems.